### PR TITLE
state db(s) refactoring using queue management

### PR DESCRIPTION
**Pull request by NikVolf**

### CODE REVIEW

File: db/src/database.rs.in [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+	}
+
+	pub fn flush_all(&self) -> Result<(), Error> {
+		let mut cache_lock = self.write_cache.write().unwrap();
+		let db_lock = self.db.read().unwrap();
+		if db_lock.is_none() { return Ok(()); }
+		let db = db_lock.as_ref().expect("we should have exited with Ok(()) on the previous step");
+
+		try!(cache_lock.flush_all(&db));
+		Ok(())
```
Comment (rphmeier) [inline db/src/database.rs.in]: could be rewritten with a match to avoid a panicker

the optimizer can have a hard time getting rid of unwind-related stuff

**Comment by NikVolf** [comment]
https://github.com/ethcore/parity/issues/313

**Comment by gavofyork** [comment]
can't merge.

any benchmarks?

**Comment by NikVolf** [comment]
tried syncing, with @arkpar also
no impact on speed syncing
probably will use less cpu when syncing blocks with many transactions

**Comment by gavofyork** [comment]
if it uses less CPU, i'd expect it to sync faster :)

this is an optimisation PR - if it doesn't give a solid benchmark improvement then there's no sense in merging.

**Comment by NikVolf** [comment]
can you suggest a benchmark that utilizes only state writes?
because sync has a bottleneck on bloom inserts right now
and this pr will also allow make a queue for bloom writes later

**Pull request closed without merging**